modules/rgbd/src/large_kinfu.cpp

```diff
@@ -0,0 +1,348 @@
+    reset();
+    submapMgr->createNewSubmap(true);
+
+}
+
+template<typename MatType>
+void LargeKinfuImpl<MatType>::reset()
+{
+    frameCounter = 0;
+    pose         = Affine3f::Identity();
+    submapMgr->reset();
+}
+
+template<typename MatType>
+LargeKinfuImpl<MatType>::~LargeKinfuImpl()
+{
+}
+
+template<typename MatType>
+const Params& LargeKinfuImpl<MatType>::getParams() const
+{
+    return params;
+}
+
+template<typename MatType>
+const Affine3f LargeKinfuImpl<MatType>::getPose() const
+{
+    return pose;
+}
+
+template<>
+bool LargeKinfuImpl<Mat>::update(InputArray _depth)
+{
+    CV_Assert(!_depth.empty() && _depth.size() == params.frameSize);
+
+    Mat depth;
+    if (_depth.isUMat())
+    {
+        _depth.copyTo(depth);
+        return updateT(depth);
+    }
+    else
+    {
+        return updateT(_depth.getMat());
+    }
+}
+
+template<>
+bool LargeKinfuImpl<UMat>::update(InputArray _depth)
+{
+    CV_Assert(!_depth.empty() && _depth.size() == params.frameSize);
+
+    UMat depth;
+    if (!_depth.isUMat())
+    {
+        _depth.copyTo(depth);
+        return updateT(depth);
+    }
+    else
+    {
+        return updateT(_depth.getUMat());
+    }
+}
+
+template<typename MatType>
+bool LargeKinfuImpl<MatType>::updateT(const MatType& _depth)
+{
+    CV_TRACE_FUNCTION();
+
+    MatType depth;
+    if (_depth.type() != DEPTH_TYPE)
+        _depth.convertTo(depth, DEPTH_TYPE);
+    else
+        depth = _depth;
+
+    std::vector<MatType> newPoints, newNormals;
+    makeFrameFromDepth(depth, newPoints, newNormals, params.intr, params.pyramidLevels, params.depthFactor,
+                       params.bilateral_sigma_depth, params.bilateral_sigma_spatial, params.bilateral_kernel_size,
+                       params.truncateThreshold);
+
+    for (const auto& it : submapMgr->activeSubmaps)
+    {
+        //! Iterate over map?
+        int currTrackingId = it.first;
+        auto submapData = it.second;
+        Ptr<Submap<MatType>> currTrackingSubmap = submapMgr->getSubmap(currTrackingId);
+        Affine3f affine;
+        std::cout << "Current tracking ID: " << currTrackingId << std::endl;
+
+        if(frameCounter == 0) //! Only one current tracking map
+        {
+            currTrackingSubmap->integrate(depth, params.depthFactor, params.intr, frameCounter);
+            currTrackingSubmap->pyrPoints = newPoints;
+            currTrackingSubmap->pyrNormals = newNormals;
+            continue;
+        }
+
+        //1. Track
+        int numInliers = 0;
+        bool trackingSuccess = false;
+        std::tie(trackingSuccess, numInliers) =
+            icp->estimateTransformInliers(affine, currTrackingSubmap->pyrPoints, currTrackingSubmap->pyrNormals, newPoints, newNormals);
+        if (trackingSuccess)
+        {
+            //! Compose current pose and add to camera trajectory
+            currTrackingSubmap->composeCameraPose(affine);
+            std::cout << "Number of inliers: " << numInliers << "\n";
+        }
```

Comment:
Since we are tracking each submap independently, we should go forward. I am yet to add some code to remove the submap from active list, if the tracking fails consistently for a few frames, etc.